src/components/ViewOnly.js

```diff
@@ -32,115 +35,164 @@ class ViewOnly extends React.Component {
+    const { mostRecentProgram, setProgramCode } = this.props;
+    const { originalCode } = this.state;
+
     if (this.savePrevProgram) {
-      this.props.setProgramCode(this.props.mostRecentProgram, this.state.originalCode);
+      setProgramCode(mostRecentProgram, originalCode);
     }
-  };
-
-  codeSaverHelper = async () => {
-    const { ok: okOriginal, sketch: original } = await fetch.getSketch(
-      this.props.mostRecentProgram,
-    );
+  }
 
-    if (!okOriginal) {
-      this.setState({ notfound: true });
-      return;
+  codeSaverHelper = () => {
+    const { mostRecentProgram } = this.props;
+
+    try {
+      fetch
+        .getSketch(mostRecentProgram)
+        .then((res) => {
+          if (!res.ok) {
+            this.setState({ notfound: true });
+            return Promise.reject(res);
+          }
+          return res.json();
+        })
+        .then((sketch) => {
+          this.setState({
+            originalCode: sketch.code,
+          });
+        })
+        .catch((err) => {
+          console.error(err);
+        });
+    } catch (err) {
+      console.error(err);
     }
-
-    this.setState({
-      originalCode: original.code,
-    });
   };
 
   onThemeChange = () => {
-    const newTheme = this.props.theme === 'dark' ? 'light' : 'dark';
+    const { theme, setTheme } = this.props;
+
+    const newTheme = theme === 'dark' ? 'light' : 'dark';
     cookies.setThemeCookie(newTheme);
-    this.props.setTheme(newTheme);
+    setTheme(newTheme);
   };
 
   render() {
-    if (this.state.notfound) {
+    const {
+      notfound, loaded, viewMode, pane1Style, language, code, sketchName, thumbnail,
+    } = this.state;
+    const {
+      left,
+      screenWidth,
+      screenHeight,
+      theme,
+      contentType,
+      panelOpen,
+      mostRecentProgram,
+      programid,
+    } = this.props;
+
+    if (notfound) {
       return <PageNotFound />;
     }
-    if (!this.state.loaded) {
+    if (!loaded) {
       return <LoadingPage />;
     }
     const codeStyle = {
-      left: this.props.left || 0,
-      width: this.props.screenWidth - (this.props.left || 0),
-      height: this.props.screenHeight,
+      left: left || 0,
+      width: screenWidth - (left || 0),
+      height: screenHeight,
     };
 
     return (
-      <div className={`main theme-${this.props.theme}`}>
+      <div className={`main theme-${theme}`}>
         <ProfilePanelContainer
-          contentType={this.props.contentType}
-          theme={this.props.theme}
+          contentType={contentType}
+          theme={theme}
           onThemeChange={this.onThemeChange}
         />
         <div className="editor" style={codeStyle}>
           <EditorAndOutput
             // view mode
-            viewMode={this.state.viewMode}
-            updateViewMode={(viewMode) => this.setState({ viewMode })}
+            viewMode={viewMode}
+            updateViewMode={(_viewMode) => this.setState({ viewMode: _viewMode })}
```

Comment:
nit: this could be named something else like `vm`, as the underscore usually implies the variable is unused